File: fraiseql-java/src/main/java/com/fraiseql/core/ObserverBuilder.java
package com.fraiseql.core;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Fluent builder for FraiseQL observer definitions.
 * Observers react to database events and dispatch configured actions.
 *
 * <p>Example:
 * <pre>
 * new ObserverBuilder("onOrderCreated")
 *     .entity("Order")
 *     .event("INSERT")
 *     .addAction(Webhook.create("https://example.com/hook"))
 *     .register();
 * </pre>
 */
public final class ObserverBuilder {

    private final String name;
    private String entity;
    private String event;
    private String condition;
    private final List<Map<String, Object>> actions = new ArrayList<>();
    private RetryConfig retry = RetryConfig.defaults();

    private static final SchemaRegistry registry = SchemaRegistry.getInstance();

    /**
     * @param name unique observer name
     */
    public ObserverBuilder(String name) {
        this.name = name;
    }

    /**
     * Set the entity type that triggers this observer.
     *
     * @param entity GraphQL type name
     * @return this builder
     */
    public ObserverBuilder entity(String entity) {
        this.entity = entity;
        return this;
    }

    /**
     * Set the database event that triggers this observer.
     *
     * @param event INSERT, UPDATE, or DELETE
     * @return this builder
     */
    public ObserverBuilder event(String event) {
        this.event = event;
        return this;
    }

    /**
     * Optional CEL-like condition; the observer fires only when true.
     *
     * @param condition filter expression
     * @return this builder
     */
    public ObserverBuilder condition(String condition) {
        this.condition = condition;
        return this;
    }

    /**
     * Add an action to execute when the observer fires.
     * Use {@link Webhook}, {@link SlackAction}, or {@link EmailAction} factories.
     *
     * @param action action map produced by an action factory
     * @return this builder
     */
    public ObserverBuilder addAction(Map<String, Object> action) {
        actions.add(action);
        return this;
    }

    /**
     * Override the default retry configuration.
     *
     * @param retry retry configuration
     * @return this builder
     */
    public ObserverBuilder retry(RetryConfig retry) {
        this.retry = retry;
        return this;
    }

    /**
     * Register this observer in the schema registry.
     *
     * @throws IllegalStateException if entity or event is missing
     */
    public void register() {
        if (entity == null || entity.isEmpty()) {
            throw new IllegalStateException("Observer '" + name + "': entity must be set");
        }
        if (event == null || event.isEmpty()) {
            throw new IllegalStateException("Observer '" + name + "': event must be set");
        }
        registry.registerObserver(name, entity, event, condition, retry, new ArrayList<>(actions));
    }
}

File: fraiseql-php/src/ObserverBuilder.php
<?php

declare(strict_types=1);

namespace FraiseQL;

/**
 * Fluent builder for creating FraiseQL observers.
 *
 * Example:
 * ```php
 * ObserverBuilder::create('onHighValueOrder')
 *     ->entity('Order')
 *     ->event('INSERT')
 *     ->condition('total > 1000')
 *     ->addAction(Webhook::create('https://api.example.com/orders'))
 *     ->addAction(SlackAction::create('#sales', 'New order: {id}'))
 *     ->retry(['max_attempts' => 5, 'backoff_strategy' => 'exponential'])
 *     ->register();
 * ```
 */
final class ObserverBuilder
{
    private string $name;
    private ?string $entity = null;
    private ?string $event = null;
    private ?string $condition = null;

    /** @var array<int, array<string, mixed>> */
    private array $actions = [];

    /** @var array<string, mixed> */
    private array $retry = [
        'max_attempts' => 3,
        'backoff_strategy' => 'exponential',
        'initial_delay_ms' => 100,
        'max_delay_ms' => 60000,
    ];

    private function __construct(string $name)
    {
        $this->name = $name;
    }

    public static function create(string $name): self
    {
        return new self($name);
    }

    public function entity(string $entity): self
    {
        $this->entity = $entity;
        return $this;
    }

    public function event(string $event): self
    {
        $this->event = strtoupper($event);
        return $this;
    }

    public function condition(string $condition): self
    {
        $this->condition = $condition;
        return $this;
    }

    /**
     * @param array<string, mixed> $action
     */
    public function addAction(array $action): self
    {
        $this->actions[] = $action;
        return $this;
    }

    /**
     * @param array<string, mixed> $retry
     */
    public function retry(array $retry): self
    {
        $this->retry = array_merge($this->retry, $retry);
        return $this;
    }

    public function register(): void
    {
        if ($this->entity === null || $this->event === null) {
            throw new FraiseQLException('Observer must have entity and event');
        }

        if (empty($this->actions)) {
            throw new FraiseQLException('Observer must have at least one action');
        }

        $definition = new ObserverDefinition(
            name: $this->name,
            entity: $this->entity,
            event: $this->event,
            actions: $this->actions,
            condition: $this->condition,
            retry: $this->retry,
        );

        SchemaRegistry::getInstance()->registerObserver($definition);
    }
}

/**
 * Helper functions for creating observer actions.
 */
final class Webhook
{
    /**
     * @param array<string, mixed> $options
     * @return array<string, mixed>
     */
    public static function create(string $url, array $options = []): array
    {
        return array_merge([
            'type' => 'webhook',
            'url' => $url,
            'headers' => ['Content-Type' => 'application/json'],
        ], $options);
    }

    /**
     * @param array<string, mixed> $options
     * @return array<string, mixed>
     */
    public static function withEnv(string $urlEnv, array $options = []): array
    {
        $action = array_merge([
            'type' => 'webhook',
            'url_env' => $urlEnv,
            'headers' => ['Content-Type' => 'application/json'],
        ], $options);

        unset($action['url']);
        return $action;
    }
}

final class SlackAction
{
    /**
     * @param array<string, mixed> $options
     * @return array<string, mixed>
     */
    public static function create(string $channel, string $message, array $options = []): array
    {
        return array_merge([
            'type' => 'slack',
            'channel' => $channel,
            'message' => $message,
            'webhook_url_env' => 'SLACK_WEBHOOK_URL',
        ], $options);
    }

    /**
     * @param array<string, mixed> $options
     * @return array<string, mixed>
     */
    public static function withWebhookUrl(string $channel, string $message, string $webhookUrl, array $options = []): array
    {
        $action = self::create($channel, $message, $options);
        $action['webhook_url'] = $webhookUrl;
        unset($action['webhook_url_env']);
        return $action;
    }
}

final class EmailAction
{
    /**
     * @param array<string, mixed> $options
     * @return array<string, mixed>
     */
    public static function create(string $to, string $subject, string $body, array $options = []): array
    {
        return array_merge([
            'type' => 'email',
            'to' => $to,
            'subject' => $subject,
            'body' => $body,
        ], $options);
    }

    /**
     * @param array<string, mixed> $options
     * @return array<string, mixed>
     */
    public static function withFrom(string $to, string $subject, string $body, string $fromEmail, array $options = []): array
    {
        return self::create($to, $subject, $body, array_merge(['from' => $fromEmail], $options));
    }
}
